### PR TITLE
Derive F30 and F3A from RMIGPIOFunction

### PR DESCRIPTION
**Pull request by zhen-zen**

### CODE REVIEW

File: VoodooRMI/Functions/F30.cpp [hunk 1]
```diff
@@ -7,95 +7,73 @@
...
-    
-    if (retval < 0)
+bool F30::init(OSDictionary *dictionary) {
+    if (!super::init(dictionary))
         return false;
-    
-    super::attach(provider);
-    
+
+    query_regs = reinterpret_cast<uint8_t *>(IOMalloc(RMI_F30_QUERY_SIZE * sizeof(uint8_t)));
```
Comment (1Revenger1) [inline VoodooRMI/Functions/F30.cpp:18]: This feels like something that should be done for all the functions or none of them. The line of what F30 should do vs the RMIGPIOFunction is a bit confusing.

I think I'll still merge it in though, if you can get the other things fixed.

**Review by 1Revenger1 — CHANGES_REQUESTED** [review]
Not everything needs to be resolved. A couple of them are just my thoughts, but I'll still pull it in if you can change `has_gpio` and init return value.

File: VoodooRMI/Functions/Input/RMIGPIOFunction.cpp [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+int RMIGPIOFunction::mapGpios()
+{
+    unsigned int button = BTN_LEFT;
+    unsigned int trackpoint_button = BTN_LEFT;
+    int buttonArrLen = min(gpioled_count, TRACKPOINT_RANGE_END);
+    const gpio_data *gpio = bus->getGPIOData();
+    setProperty("Button Count", buttonArrLen, 32);
+
+    for (int i = 0; i < buttonArrLen; i++) {
+        if (!is_valid_button(i))
```
[the remark this answers is not in the document]
Comment (zhen-zen) [reply VoodooRMI/Functions/Input/RMIGPIOFunction.cpp:64]: Maybe it's better to have them in a separate commit so we can track changes easier.

File: VoodooRMI/Functions/Input/RMIGPIOFunction.cpp [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+bool RMIGPIOFunction::attach(IOService *provider)
+{
+    int error;
+
+    error = initialize();
+    if (error)
+        return false;
+
+    super::attach(provider);
```
[the remark this answers is not in the document]
Comment (zhen-zen) [reply VoodooRMI/Functions/Input/RMIGPIOFunction.cpp:22]: Actually I'm confused on this. `attach` is used to call `rmi_f30_initialize` so I kept the origin structure. I was considering to move some common parts to `attach`, i.e. initial query and stuffs after `readControlParameters`.

File: VoodooRMI/Functions/F30.cpp (hunk 1 above)
Comment (zhen-zen) [reply VoodooRMI/Functions/F30.cpp:18]: I want to replace the default size to a dynamic one but haven't come up with a good idea.

Comment (zhen-zen) [reply VoodooRMI/Functions/F30.cpp]: How about the last commit?

File: VoodooRMI/Functions/F30.cpp [hunk 4]
```diff
@@ -55,12 +29,20 @@ int F30::initialize()
...
+    }
+    memset(query_regs, 0, query_regs_size * sizeof(uint8_t));
+
     error = bus->readBlock(desc.query_base_addr,
                               query_regs, RMI_F30_QUERY_SIZE);
     if (error) {
-        IOLogError("F30: Failed to read query register");
-        return error;
+        IOLogError("%s: Failed to read query register: %d", getName(), error);
+        return -1;
```
Comment (zhen-zen) [inline VoodooRMI/Functions/F30.cpp:45]: Typo. To be fixed in the next commit.

File: VoodooRMI/Functions/Input/RMIGPIOFunction.hpp [hunk 5]
```diff
@@ -18,22 +18,23 @@ class RMIGPIOFunction : public RMIFunction {
...
     RelativePointerEvent relativeEvent {};
 
     uint8_t *query_regs {nullptr};
-//    uint32_t query_regs_size {0};
+    uint32_t query_regs_size {0};
     uint8_t *ctrl_regs {nullptr};
     uint32_t ctrl_regs_size {0};
     uint8_t *data_regs {nullptr};
-//    uint32_t data_regs_size {0};
+    uint32_t data_regs_size {0};
```
Comment (zhen-zen) [inline VoodooRMI/Functions/Input/RMIGPIOFunction.hpp:31]: Actually it's just `register_count`?

File: VoodooRMI/Functions/Input/RMIGPIOFunction.cpp [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+    return true;
+}
+
+int RMIGPIOFunction::config()
```
Comment (zhen-zen) [inline VoodooRMI/Functions/Input/RMIGPIOFunction.cpp:27]: It seems that `F3A` doesn't do this step?

Comment (1Revenger1) [reply VoodooRMI/Functions/Input/RMIGPIOFunction.cpp:27]: No, probably not needed in F03 either. Mostly if we decide to do any configuration like for LEDs. That's mostly an F03 thing though.

File: VoodooRMI/Functions/F30.cpp [hunk 7]
```diff
@@ -161,23 +138,24 @@ int F30::rmi_f30_initialize()
     ctrl_regs_size = (uint32_t) (ctrl_reg -
```
Comment (zhen-zen) [inline VoodooRMI/Functions/F30.cpp:138]: ctrl_regs_size might be shrinked here and missed during free()

File: VoodooRMI/Functions/F30.cpp (hunk 1 above)
Comment (1Revenger1) [reply VoodooRMI/Functions/F30.cpp]: F30 is pretty complicated. It's got a lot of ctrl registers for haptics and LEDs that I eventually want to mess around with. I think trying to do any sort of inheritance for the registers is going to end up more complicated in the end. I think an interface instead where F30/F3A loops through the each GPIO pin, figures out if it's valid, and then registers it with RMIGPIOFunction might work better? That way the query/ctrl registers stay in F30/F3A and we don't need to alloc anything. Still need to allocate the button array, but we might be able to use a set method in RMIGPIOFunction right before we start registering buttons? If you've got another solution, I'd be happy to hear it. I don't think this is quite it though.

**Comment by zhen-zen** [comment]
Sorry for the late update. Just came back from vacation.

**Pull request merged**